kernel-rs/src/virtio/virtio_disk.rs

```diff
@@ -352,6 +375,210 @@ impl VirtioDisk {
+                                guard.sleep(ctx);
+                                continue;
+                            }
+                            idx => header = idx,
+                        }
+                        match guard.get_pin_mut().alloc() {
+                            None => {
+                                guard.get_pin_mut().as_mut().free(header.unwrap());
+                                header = None;
+                                continue;
+                            }
+                            idx => tailer = idx,
+                        }
+                    }
+
+                    // Allocate a descriptor for the buffer.
+                    match guard.get_pin_mut().alloc() {
+                        Some(idx) => break idx,
+                        // Back down to prevent deadlock
+                        None => {
+                            if Self::finalize_write_seq(
+                                guard,
+                                &mut darray[next_write..=prev_blk_idx],
+                                &mut barray[next_write..=prev_blk_idx],
+                                &mut header,
+                                &mut tailer,
+                                ctx,
+                            ) {
+                                next_write = prev_blk_idx + 1;
+                            } else {
+                                // Waiting for other threads to free the descriptors.
+                                guard.sleep(ctx);
+                            }
+                        }
+                    };
+                };
+
+                let buf = &mut barray[i].as_mut().unwrap();
+
+                // Set the descriptor of the buffer
+                let this = guard.get_pin_mut().project();
+                this.desc[desc.idx] = VirtqDesc {
+                    addr: buf.deref_inner().data.as_ptr() as _,
+                    len: BSIZE as _,
+                    flags: VirtqDescFlags::NEXT,
+                    next: 0,
+                };
+
+                // Chain with the previous descriptor
+                if let Some(prev_d) = &darray[prev_blk_idx] {
+                    this.desc[prev_d.idx].next = desc.idx as _;
+                }
+
+                buf.deref_inner_mut().disk = true;
+
+                darray[i] = Some(desc);
+                prev_blk_idx = i;
+            }
+        }
+
+        let _ = Self::finalize_write_seq(
+            guard,
+            &mut darray[next_write..=prev_blk_idx],
+            &mut barray[next_write..=prev_blk_idx],
+            &mut header,
+            &mut tailer,
+            ctx,
+        );
+    }
+
+    // This method writes the buffers in the given range of the given array to the disk.
+    fn finalize_write_seq(
+        guard: &mut SleepableLockGuard<'_, Self>,
+        darray: &mut [Option<Descriptor>],
+        barray: &mut [Option<Buf>],
+        header: &mut Option<Descriptor>,
```

Comment:
There are times when `header` and tailer` are not allocated, e.g. after calling `Self::finalize_wrrite_seq`, so I have to use `Option` here. An alternative way is to make `Self::finalize_write_seq` allocate and return a new header and tailer descriptor every time it is called.

Reply:
I think we shouldn't try to call `Self::finalize_write_seq` when we didn't allocate a descriptor for `header` and `tailer`. We should force the caller to do this by asking for a `&mut Descriptor` instead.

Reply:
Since we have to free the descriptors after they are used by `Self::finalize_write_seq` and `VirtioDisk::free` takes ownership of `Descriptor`. Do you also think that we should free the descriptors in the caller method instead?

Reply:
* Oh, then maybe we should just make `header: Descriptor` instead and free it inside `Self::finalize_write_seq`.
* However, as I said before, I don't think this is really important.